astropy/io/fits/tests/test_diff.py

```diff
@@ -554,7 +554,7 @@ def test_different_table_data(self):
         assert ('Column A data differs in row 0:\n'
                 '    a> True\n'
                 '    b> False') in report
-        assert ('...and at 13 more indices.\n'
+        assert ('...and at 1 more indices.\n'
```

Comment:
I am a bit concerned in the change of this number. `13 -> 1` does not seem correct here given that "13" is still tested for in two other places in this same test. Perhaps we should take a step back and modify the code in such a way that it does not fundamentally change that `fitsdiff` does...

Reply:
This change can be explained: this is an array with 4 elements (`2x2`) and so there are 4 differing elements. In the past `fitsdiff` thought that a `2x2` array would have 16 different elements. `fitsdiff` displays first three differences so that in the new version (that correctly, _IMO_, computes 4 elements) there is "1 more" different element (that was not displayed) while in the old version (the one that was computing 16 differing elements for a `2x2` array) after displaying first three different elements there were "13 more" left.

I do not thing we need to take a step back. Instead, I would challenge you to display all 16 different elements in the old code. That is, in `islice` do not limit the code to take the first three elements - print them all and see if you get 16 different elements.

Reply:
The other 13 in the test comes form the fact that there are `fitsdiff` has no idea how to count things (I left that one for you @pllim :-) - well it needs some discussion/understanding of what was the intension). Take a look at the "report":

Reply:
... or, is this just a bad test setup?! Column C is the only one that has 4 "rows" (array data of shape `2x4`) while all other columns have only two rows (array data of shape Nx2)!

Reply:
... or, maybe, it's not a bad test but just how things are supposed to be for this particular `dim` parameter. I just do not know what is going on here. In any case, the two 13th reported in the original test are unrelated.

Reply:
> Let's consider "Column G" for example. This one contains this array array=[0.0, 1.0] (which has a single row (in my interpretation) with two columns)

No, it's a column, with 2 rows.

Maybe easier to see with the table:
```
In [8]: ta.data
Out[8]: 
FITS_rec([( 1, [  0], [[0, 1], [2, 3]], 0, 'abc', 0., 0., 0.+1.j, 4.+5.j, [0, 1]),
          ( 0, [  1], [[4, 5], [6, 7]], 2, 'def', 1., 1., 2.+3.j, 6.+7.j, [2, 3])],
         dtype=(numpy.record, [('A', 'i1'), ('B', 'u1', (1,)), ('C', '<i2', (2, 2)), ('D', '<i4'), ('E', 'S3'), ('F', '<f4'), ('G', '<f8'), ('H', '<c8'), ('I', '<c16'), ('J', '<i4', (2,))]))

In [9]: Table(ta.data)
Out[9]: 
<Table length=2>
  A   B [1] C [2,2]   D    E      F       G        H         I        J   
 bool  bool  int16  int32 str3 float32 float64 complex64 complex128 object
----- ----- ------- ----- ---- ------- ------- --------- ---------- ------
 True False  0 .. 3     0  abc     0.0     0.0        1j     (4+5j)  [0 1]
False  True  4 .. 7     2  def     1.0     1.0    (2+3j)     (6+7j)  [2 3]
```

> For C there is a single row with 4 columns in the array (array=[[0, 1, 2, 3], [4, 5, 6, 7]])

Because of `dim`, C is a special beast here. Each row element is a 2x2 array.

So, in my understanding the report looks correct (once you know what's in the table).

Reply:
@saimn Thanks for clarifying this! I think I now understand what is going on with C. In this case the reported 13 differences in the table look correct and, IMO, the reported "... 1 more" for the row 1 in column C also seems correct. I think we should not revert to the old logic of counting differences by which a `5x5` array can have 625 differences.

Reply:
Agreed, I don't understand either the old `num_diffs` logic, and the [history](https://github.com/spacetelescope/PyFITS/commit/8e4d6d20e0fddba232f88ce3edf985b871816910#diff-ec63a1f7dd27e3295d30eb67aabc6e79L1080) of it does not help much.

Reply:
So, what does this mean? There needs to be two different logic when dealing with columns and when dealing with Numpy arrays?

Reply:
On the contrary, now in both cases the same logic of counting differences is used. That is, now when counting arrays you get the _actual_ number of differences - not some unjustifiable number.

Reply:
Yes. I think you are right. I'll need to wrap my head around this but not right now...

Reply:
Yes, the old formatting for this absolutely sucked.  The use of the diff format is useful for string values, but not for numerical differences.